Adds setting to exclude artifacts in link metadata

lukpueh opened the pull request:
- Adds routine to exclude files, matched by defined patterns (uses fnmatch), when recording artifacts (materials or products) upon in-toto-run/in-toto-record
- For now the exclude patterns can be defined as a setting ARTIFACT_EXCLUDES. This is likely to become a command line arg or part of a config file, e.g. .in-toto-ignore
- Set ARTIFACT_EXLCUDES to ignore ".git" and link metadata files
- Adds unit tests

Commit:
Adds setting to exclude artifacts in link metadata

- Adds routine to exclude files, matched by defined patterns
 (uses fnmatch), when recording artifacts (materials or products)
 upon in-toto-run/in-toto-record

- For now the exclude patterns can be defined as a setting
  ARTIFACT_EXCLUDES. This is likely to become a command line arg
  or part of a config file, e.g. .in-toto-ignore

- Set ARTIFACT_EXLCUDES to ignore ".git" and link metadata files

- Adds unit tests

Commit:
Adds apply exclude patterns to mid-level dirs

Exclude patterns need to be applied three times:
(1) On the first level of the directory tree when artifacts could
be either files or directories,
(2) on intermediate directories, i.e. dirs that are not on the first
level, and
(3) on tree leaves that are not on the first level.

EXAMPLE:
.
├── foo.pyc
└── sub
    ├── .git
    └── bar.pyc
ARTIFACT_EXCLUDES = [".git", "*.pyc]
record_artifacts_as_dict(".")

(1) excludes foo.pyc
(2) excludes sub/.git
(3) excludes bar.pyc

This commit adds (2) + unit tests

lukpueh commented:
Exclude patterns are applied 3 times:
(1) On the first level of the directory tree when artifacts could be either files or directories,
(2) on intermediate directories, i.e. dirs that are not on the first level, and
(3) on tree leaves that are not on the first level.

*EXAMPLE:*
```
# Directory Tree
.
├── foo.pyc
└── sub
    ├── .git
    └── bar.pyc

# Setting
ARTIFACT_EXCLUDES = [".git", "*.pyc]
# Call record_artifacts_as_dict to record from root of above tree
record_artifacts_as_dict(".")
```
(1) excludes foo.pyc
(2) excludes sub/.git
(3) excludes bar.pyc

SantiagoTorres commented on test/test_runlib.py:
```diff
@@ -22,24 +22,126 @@
+
+    self.assertListEqual(sorted(artifacts_dict.keys()),
+        sorted(["foo", "bar", "subdir/foosub"]))
+    settings.ARTIFACT_EXCLUDES = []
```
I wonder if it would be a good idea to have this under a `try:finally:` or a `TearDown()` so that clearing the excludes is always performed and doesn't pollute ensuing tests.

lukpueh replied:
Excellent idea. Will do.

SantiagoTorres commented on test/test_runlib.py:
```diff
@@ -22,24 +22,126 @@
+    os.mkdir("subdir")
+    os.mkdir("subdir/subsubdir")
+    open("subdir/foosub", "w").write("foosub")
+    open("subdir/subsubdir/foosubsub", "w").write("foosubsub")
```
I'm curious, could we add a file that hashes to the same (e.g,. a second file with "foosub") to trigger the "files with the same hash" cornercase?

lukpueh replied:
Yes and no. This TestCase only checks `runlib.record_artifacts_as_dict` which recursively traverses the passed list of paths and stores each file together with its hash. It does not care for hash collisions.

The "files with the same hash" corner case only appears in `verifylib.verify_match_rule`, where we try to map files from one artifact dict (as recorded above) to the files from another artifact dict.

Moreover, the hash collision issue has been dealt with in: https://github.com/in-toto/in-toto/commit/75ca2ce6f4b04fb9ca3bda66b282371883de9eb1
which I used for the [in-toto in Seattle](https://github.com/in-toto/in-toto/tree/seattle) show case, but which has not been PRed yet, because of ongoing discussions in https://github.com/in-toto/in-toto/issues/43.

SantiagoTorres replied:
Hmm, I still think it wouldn't hurt to have this to avoid any sort of "regression"? (I know it's not a regression, but oh well)

lukpueh replied:
Okay, I'll add it since it does not hurt as you say.

SantiagoTorres commented on test/test_runlib.py:
```diff
@@ -22,24 +22,126 @@
+  @classmethod
+  def tearDownClass(self):
+    """Change back to initial working dir and remove temp test directory. """
+    os.chdir(WORKING_DIR)
```
I'm thinking that maybe this global WORKING_DIR could be part of the object and populated upon `setUpClass()`. Just for the sake of symmetry with `self.test_dir`

lukpueh replied:
Yes, makes sense. My reasoning was to reduce some redundant code, because there are multiple test cases that work in a similar way in terms of a temp directory. But I guess the global is only a half-baked approach. I'll move the global to the setUpClass methods.

SantiagoTorres commented on test/test_runlib.py:
```diff
@@ -22,24 +22,126 @@
 from in_toto.util import generate_and_write_rsa_keypair, \
     prompt_import_rsa_key_from_file
 from in_toto.models.link import Link
 from in_toto.exceptions import SignatureVerificationError
```
Is this used anywhere?

SantiagoTorres replied:
Same for `Link`

lukpueh replied:
Yes. :)

SantiagoTorres replied:
kk, I was just making sure. I grepped for it and it's indeed used.

SantiagoTorres commented on in_toto/settings.py:
```diff
@@ -65,6 +65,8 @@
 GENERAL_CRYPTO_LIBRARY = 'pycrypto'
 
+# FIXME: Add as command line argument or to config file, e.g .in-toto-ignore
+ARTIFACT_EXCLUDES=["*.link*", ".git"]
```
I like these defaults. I wonder if we can extend this list with `*.pyc` and  `*~`

SantiagoTorres commented on in_toto/runlib.py:
```diff
@@ -82,6 +92,14 @@ def record_artifacts_as_dict(artifacts):
+    Excludes files that are matched by the file patterns specified in
+    ARTIFACT_EXCLUDES setting.
+    Note: Exclude patterns are applied on the basename, i.e. the pattern
+    "foo" excludes /foo as well as /subdir/foo
```
I see that this is a consequence of using fnmatch. Maybe in the future we want to change this behavior? (e.g,. have `**/.git` to ignore all .git folders in any subdirectory and `.git` to only ignore the one in root).

lukpueh replied:
Actually it is not a consequence of using fnmatch. I implemented this behavior on purpose, because I thought this is how .gitignore works. But maybe I was wrong. Do you want me to change this now?

SantiagoTorres replied:
Hmm, I'm assuming it would be a good idea to mimic the behavior/semantic that .gitignore does to simplify adoption. Both files, gitignore and in-totoignore, serve the same purpose, so why not?

We could ticketize it and move forward with the PR though.

lukpueh replied:
You know what. I'll add it right away. Let me revise the .gitignore behavoir...

SantiagoTorres reviewed (commented):
I think none of these issues is a blocker, but I'd like to see what's your insight on my comments before I merge. 👍

Commit:
Moves WORKING_DIR to setUpClass in test_runlib

Remove the global WORKING_DIR and replace it with attributes
self.working_dir in relevant setUpClass methods.
While this introduces some redundancy it makes the code clearer.

Commit:
Clears artifacts_excludes in test_runlib tearDown

Commit:
Adds "*.pyc" and "*~" to default exclude settings

Commit:
Adds a same-hash artifact to RecordArtifacts test

Hash collisions should not affect the way artifacts are recorded.
They are only relevant when mapping files from one artifact
dict to another, e.g. while using match rules.

We add two artifacts with the same hash to the record artifacts
TestCase to demonstrate that it does not have an effect.

Commit:
Replaces normalize path, updates exclude pattern

- Replaces custom _normalize path with os.path.normpath
- Exclude patterns are applied on normalized paths
- Normalized paths are stored as keys for the  artifact_dict
- Exclude patterns are now applied on the full paths and not
  only on the basenames

lukpueh commented on in_toto/runlib.py:
```diff
@@ -82,6 +92,14 @@ def record_artifacts_as_dict(artifacts):
+    Excludes files that are matched by the file patterns specified in
+    ARTIFACT_EXCLUDES setting.
+    Note: Exclude patterns are applied on the basename, i.e. the pattern
+    "foo" excludes /foo as well as /subdir/foo
```
sdf

lukpueh commented:
Updated in c062460:

*EXAMPLE:*
```
# Directory Tree
.
├── foo.pyc
└── sub
    ├── .git
    └── bar.pyc

# Setting
ARTIFACT_EXCLUDES = [".git", "*.pyc]
# Call record_artifacts_as_dict to record from root of above tree
record_artifacts_as_dict(".")
```
(1) excludes foo.pyc
(2) ~~excludes sub/.git~~ (`"*.git"` would also exclude `sub/.git`)
(3) excludes bar.pyc

SantiagoTorres commented on in_toto/runlib.py:
```diff
@@ -119,28 +133,56 @@ def record_artifacts_as_dict(artifacts):
+
+        # Reset and refill dirs with remaining names after exclusion
+        # Modify (not reassign) dirnames to only recurse into remaining dirs
+        dirs[:] = []
```
hmm, I wonder if this is the behavior that we want. You are assigning the same list to all the elements in dir... Have you verified this behavior?

lukpueh replied:
`dirs[:] = []` is equal to `del dirs[:]` is equal to `clear the list` which is the behavior I want

lukpueh replied:
Modifying the `dirs` list prevents `os.walk` from further recursing into directories that are no longer in the list.

lukpueh replied:
From the python [`os.walk docs`](https://docs.python.org/2/library/os.html#os.walk)
> ...the caller can modify the dirnames list in-place (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames...

SantiagoTorres replied:
Right, I just verified this behavior. Merging!
